searchState.loadedDescShard("hidive", 0, "Hidive is a targeted genome co-assembler for biobank-scale …\nAssemble target locus from long-read data in anchor-based …\nBuild series-parallel graph from long-read data in …\nCluster sequences based on k-mer presence/absence.\nCo-assemble target locus from long-read and short-read …\nStream selected loci from FASTA and long-read WGS BAM …\nFilter rescued reads to those most closely matching the …\nCluster edge matrix and impute missing edges.\nFind more sequences (aligned or unaligned) overlapping …\nTrain a graph-cleaning model using short- and long-read …\nTrim reads to a specific window around locus.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nMulti-sample BAM file with reads spanning locus of …\nTurn on debug mode.\nMulti-sample FASTA file with reads spanning locus of …\nMulti-sample FASTA file with reads spanning locus of …\nFASTA files with reads to use as a filter for finding more …\nSeries-parallel graph.\nSeries-parallel graph.\nNumber of training iterations.\nKmer-size.\nKmer-size.\nKmer-size\nKmer-size\nKmer-size\nKmer-size\nZero or more genomic loci (“contig:start-stop”) to …\nOne or more genomic loci (“contig:start-stop”) to …\nFASTA files with long-read sequences (may contain one or …\nFASTA files with long-read sequences (may contain one or …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nMinimum number of k-mers to require before examining a …\nMinimum percentage of bases in short-read sequences …\nTrained error-cleaning model.\nOutput path for trained model.\nOutput path for FASTA file with reads spanning locus of …\nOutput path for FASTA file with reads spanning locus of …\nOutput path for filtered short-read sequences.\nOutput path for clustered sequences.\nOutput path for trimmed BAM.\nOutput path for series-parallel graph.\nOutput path for series-parallel graph with imputed missing …\nOutput path for assembled long-read sequences.\nOutput path for assembled short-read sequences.\nName of sequence to use as reference.\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nFASTA files with short-read sequences (may contain one or …\nFASTA files with short-read sequences (may contain one or …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nIndexed BAM files to use as ground truth (usually from …\nInclude unmapped reads.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nStarts the data fetching process.")