searchState.loadedDescShard("hidive", 0, "Hidive is a targeted genome co-assembler for biobank-scale …\nError-correct long reads using a linked de Bruijn graph.\nBuild series-parallel graph from long-read data in …\nCall and phase variants.\nCluster sequences based on k-mer presence/absence.\nCo-assemble target locus from long- and short-read data …\nError-clean long reads using a linked de Bruijn graph.\nEvaluate a trained model using long- and (optionally) …\nStream loci from CRAM/BAM/FASTA files stored locally or in …\nFurther filter rescued reads to those most closely …\nCluster edge matrix and impute missing edges.\nFrom rescued reads, recruit subset reads that are similar …\nFind more sequences (both aligned and unaligned) …\nTrain a error correction model with reads and ground truth …\nTrim reads to a specific window around locus.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nMulti-sample BAM file with reads spanning locus of …\nBAM file with integrated long- and short-read data.\nTurn on debug mode.\nMulti-sample FASTA file with reads spanning locus of …\nMulti-sample FASTA file with reads spanning locus of …\nFASTA files with reads to use as a filter for finding more …\nFASTA files with reads to use as a filter for finding more …\nOutput path for corrected reads.\nFASTA files with short-read sequences (may contain one or …\nSeries-parallel graph.\nNumber of training iterations.\nJaccard threshold\nKmer-size.\nKmer-size.\nKmer-size.\nKmer-size.\nKmer-size\nKmer-size\nKmer-size\nKmer-size\nKmer-size\nOne or more genomic loci (“contig:start-stop[|name]”, …\nOne or more genomic loci (“contig:start-stop[|name]”, …\nOne or more genomic loci (“contig:start-stop”) to …\nOne or more genomic loci (“contig:start-stop[|name]”, …\nOne or more genomic loci (“contig:start-stop[|name]”, …\nFASTA files with long-read sequences (may contain one or …\nFASTA files with long-read sequences (may contain one or …\nFASTA files with long-read sequences (may contain one or …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nMinimum percentage of k-mers to require before examining a …\nMinimum percentage of k-mers to require before examining a …\nPath to trained model.\nTrained error-cleaning model.\nTrained error-cleaning model.\nTrained error-cleaning model.\nOutput path for trained model.\nOutput path for trained model.\nOutput path for FASTA file with reads spanning locus of …\nOutput path for FASTA file with reads spanning locus of …\nOutput path for FASTA file with reads spanning locus of …\nOutput path for filtered short-read sequences.\nOutput path for clustered sequences.\nOutput path for trimmed BAM.\nOutput path for series-parallel graph.\nOutput path for series-parallel graph with imputed missing …\nOutput path for corrected reads.\nOutput path for corrected reads.\nOutput path for assembled short-read sequences.\nOutput path for assembled short-read sequences.\nInclude unmapped reads.\nReference FASTA (for guessing where reads mapped based on …\nFASTA reference sequence.\nFASTA files with reference subsequences.\nName of sequence to use as reference.\nSample name.\nOption to search for reads based on alignment status or …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nFASTA files from which to extract relevant sequences.\nFASTA files with short-read sequences (may contain one or …\nFASTA files with short-read sequences (may contain one or …\nFASTA files with short-read sequences (may contain one or …\nFASTA files with short-read sequences (may contain one or …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nIndexed BAM files to use as ground truth (usually from …\nIndexed BAM files to use as ground truth (usually from …\nIndexed BAM files to use as ground truth (usually from …\nIndexed BAM files to use as ground truth (usually from …\nIndexed BAM files to use as ground truth (usually from …\nWindow size\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nStarts the data fetching process.\nThis function collects the loci from the mappings.\nThis function creates a fetch definition.\nThis function detects the relevant loci in the reference …\nThis function processes all or unmapped reads.\nThis function finalizes the progress.\nReturns the argument unchanged.\nThis function initializes the aligner.\nCalls <code>U::from(self)</code>.\nThis function checks if a read is valid.\nThis function maps the sequences to the reference genome …\nThis function merges overlapping intervals.\nThis function will populate the fetches with the input …\nThis function populates the interval trees with the loci.\nThis function prepares the fetches for searching. If the …\nThis function will print the fetches’ info.\nThis function retrieves the records from the reader.\nThis function updates the found progress.\nThis function updates the processed progress.\nComputes FPR and TPR at various thresholds.\nComputes precision, recall, and F1 score on test data.\nCreates a dataset for the ML model.")