searchState.loadedDescShard("hidive", 0, "Hidive is a targeted genome co-assembler for biobank-scale …\nAssemble target locus from long-read data in …\nBuild series-parallel graph from long-read data in …\nCluster sequences based on k-mer presence/absence.\nCo-assemble target locus from long-read and short-read …\nStream selected loci from FASTA and long-read WGS BAM …\nCluster edge matrix and impute missing edges.\nTrim reads to a specific window around locus.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nMulti-sample BAM file with reads spanning locus of …\nMulti-sample FASTA file with reads spanning locus of …\nMulti-sample FASTA file with reads spanning locus of …\nSeries-parallel graph.\nSeries-parallel graph.\nKmer-size\nKmer-size\nKmer-size\nOne or more genomic loci (“contig:start-stop”) to …\nOne or more genomic loci (“contig:start-stop”) to …\nFASTA files with long-read sequences (may contain one or …\nOutput path for multi-sample BAM file with reads spanning …\nOutput path for clustered sequences.\nOutput path for trimmed BAM.\nOutput path for series-parallel graph.\nOutput path for series-parallel graph with imputed missing …\nOutput path for assembled long-read sequences.\nOutput path for assembled short-read sequences.\nName of sequence to use as reference.\nIndexed WGS BAM, CRAM, or FASTA files from which to …\nFASTA files with short-read sequences (may contain one or …\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nStarts the data fetching process.")